### DISCUSSION TITLE
Feature request: option in `Meta` to rename fields in `ModelSerializer`

### DISCUSSION THREAD

**Discussion by christophehenry** [encode/django-rest-framework#8994] [Ideas & Suggestions] [answered]
As far as I know, there are 3 ways to serialize a Django model field under a different name:

- use `serializers.ModelField`
- use `serializers.SerializerMethodField`
- use the `source` parameter of `fields.Field`.

All of them are quite tedious. The first one necessitate to intantiate a least one instance of the model, which is a waste of resources and rather counter intuitive, the second necessitates to create accessors for every field and the third defeats the auto-mapping feature of `ModelSerializer`.

It would be nice if renaming a field was simpler. It could take the form of a `renamed_fields` attribute in `ModelSerializer.Meta`. Alternatively, `ModelField` can be modified to accept a `model_field_name` argument and fetch the model field based on this (although I have a preference for the first method).

I am ready to implement that feature with a little guidance.

**Comment by auvipy** [comment]
can you share some code sample/example of what you are suggesting?

**Reply by christophehenry** [reply]
Ok, let's take the following django model:

```python
from django.db import models


class Connection(models.Model):
    date = models.DateTimeField(auto_now_add=True)
    user = organisation = models.ForeignKey(models.CASCADE)
```

And I'd like it serialized like this:

```json
{
    "creation_date": "…"
    "user": "…"
}
```

In this case, I just want the `date` field renamed `creation_date`. 

Currently, one of the solution I have is:

```python
from rest_framework import serializers


class OrganisationSerializer(serializers.HyperlinkedModelSerializer):
    creation_date = serializers.ModelField(model_field=Connection()._meta.get_field("date"))

    class Meta:
        model = Connection
        fields = ["creation_date", "user"]
```

For simple renames like this, it would be nice to have something like this instead:


```python
from rest_framework import serializers


class OrganisationSerializer(serializers.HyperlinkedModelSerializer):
    class Meta:
        model = Connection
        fields = ["creation_date", "user"]
        renamed_fields = {"creation_date": "date"}
```

**Reply by auvipy** [reply]
I think you can simply rename it in the fields meta. so no other extra work or extra field attribute is not needed. you can easily change to fields = ["creation_date", "user"] from fields = ["date", "user"]

**Reply by christophehenry** [reply]
But that's not what I want to do. What I would like is, for instance `Connection.date` be serialized as `creation_date` using a simple declaration in `Meta`.

**Comment by awbacker** [comment]
A feature like this would be useful, so I could rename the output field.  Currently you need to add a new field by that name on the serializer, and specify all fields explicitly.

I find myself wanting all the auto-generated field properties (max length, help text, etc) but _only_ rename the field.  I suppose an alternative would be to make creating a serializer field from a model field easier.

Perhaps something like this:

```py
class User(Model):
    name = CharField()
    email_address = CharField(max_length=1000)

class UserSerializer(ModelSerializer):
    class Meta:
        model = User
        fields = ["name", "email"] # or "__all__"
        renamed_fields = { 
            "email": "email_address"
        }
```

If I could easily create a serializer field from a model field that would work as well. I always specify all my fields directly anyway

```py
class UserSerializer(ModelSerializer):
    email = ModelSerializer.build_field(User.email_address)

    class Meta:
        model = User
        fields = ["name", "email"]
```

**Comment by ahosie** [comment, marked as answer]
I've smacked into this one today and the fix is actually pretty straightforward and django-esque.

Set a `source` attribute and stick to the [regular field serializers](https://github.com/encode/django-rest-framework/blob/a8595a8/rest_framework/serializers.py#L905) wherever you can. Using ModelField should probably be an exception not the rule.

Given the posted example:

```python
from django.db import models

class Connection(models.Model):
    date = models.DateTimeField(auto_now_add=True)
    user = organisation = models.ForeignKey(models.CASCADE)
```

Then a corresponding serializer that does what the OP wants would be:

```python
from rest_framework import serializers

class OrganisationSerializer(serializers.HyperlinkedModelSerializer):
    creation_date = serializers.DateTimeField(source="date")

    class Meta:
        model = Connection
        fields = ["creation_date", "user"]
```
